tidy3d/components/medium.py

```diff
@@ -93,6 +165,22 @@ class AbstractMedium(ABC, Tidy3dBaseModel):
         "useful in some cases.",
     )
 
+    nonlinear_spec: NonlinearSpecType = pd.Field(
+        None,
+        title="Nonlinear Spec",
+        description="Nonlinear spec applied on top of the base medium properties.",
+    )
+
+    @pd.validator("nonlinear_spec", always=True)
```

Comment:
feel like we need a better way to do this more generally. so let's say we have 
```py
NonlinearSpecType = Union[None, Spec1, Spec2, ..., SpecN]
```
and then classes `ClassA`, `ClassB`, ... that each can only accept a certain subset of `NonlinearSpecType`.

Maybe the most general way is to just define in the `AbstractMedium`
```py
nonlinear_spec : None = pd.Field(None, ...)
```
and then in the individual classes
```py
nonlinear_spec: Union[None, Spec1]
```
? Which maybe becomes a bit annoying as the combinations grow but is reasonably general? Would pydantic take care of erroring if one tried to put the wrong spec in?  Basically I feel it might be nice to do this only in the type annotation rather than the validator. This also makes it more clear in the docs.

Reply:
On the other hand, if everything is merged into the same medium, it might be harder to keep track of the dependencies just in the type annotation. Say a non-dispersive medium can accept one type of NonlinearSpec but a dispersive can only accept another. There would be some cross coupling between the two specs that I think could only be handled through validators, so not sure. Any thoughts?

Reply:
My opinion is that the type annotations are not the best tool for this. The incompatibilities seem to usually be from things that haven't been implemented yet, and cross coupling between different specs (for example, once pole residue is a spec, nonlinear + dispersive)

Reply:
If we did go with the type annotation approach, like Daniil mentioned, we would still want validators that would give more helpful error messages in certain cases, like when Chi3Spec is added to CustomIsotropicMedium. So my feeling is that the type annotations add boilerplate without much content

Reply:
but the real argument is actually like you said, the cross coupling between different specs down the line.

Reply:
true. If we had a set of "allowed" combinations, though. we could always have a generic validator like
```
def validate_specs(cls, values):
    nonlinear_spec_type = type(values.get("nonlinear_spec"))
    dispersion_spec_type = type(values.get("dispersion_spec"))
    if (nonlinear_spec_type, dispersion_spec_type) not in ALLOWED_SPEC_COMBINATIONS:
         raise ValidationError(f"Combination of {nonlinear_spec_type} and {dispersion_spec_type} not currently supported.").
```
which could be cleaner and we could document the `ALLOWED_SPEC_COMBINATIONS`. What I worry about slightly is a set of validators with really convoluted logic and no simple way for the user to see what combinations are actually allowed. However, I guess if the allowed combinations is pretty easy to write down as a function, it might indeed be easier? @dbochkov-flexcompute any thoughts?